test/jdk/java/util/concurrent/Executors/AutoShutdown.java

```diff
@@ -23,89 +23,115 @@
+        ExecutorService executor = supplier.get();
+        // submit a task to get a worker to start
+        executor.submit(() -> null).get();
+        ExecutorService delegate = getDelegate(executor);
+        executor = null;
+        gcAndAwaitTermination(delegate);
+    }
+
+    /**
+     * SingleThreadExecutor with an active worker and queued tasks.
+     */
+    @ParameterizedTest
+    @MethodSource("executorAndQueuedTaskCounts")
+    void testActiveWorker(Supplier<ExecutorService> supplier,int queuedTaskCount) throws Exception {
+        ExecutorService executor = supplier.get();
+        // the worker will execute one task, the other tasks will be queued
+        int ntasks = 1 + queuedTaskCount;
+        AtomicInteger completedTaskCount = new AtomicInteger();
+        for (int i = 0; i < ntasks; i++) {
+            executor.submit(() -> {
+                Thread.sleep(Duration.ofMillis(500));
+                completedTaskCount.incrementAndGet();
+                return null;
+            });
         }
-        if (!done)
-            throw new AssertionError("pool threads did not terminate");
+        ExecutorService delegate = getDelegate(executor);
+        executor = null;
+        gcAndAwaitTermination(delegate);
+        assertEquals(ntasks, completedTaskCount.get());
+    }
+
+    /**
+     * Returns the delegates for the given ExecutorService. The given ExectorService
+     * must be a Executors$DelegatedExecutorService.
+     */
+    private ExecutorService getDelegate(ExecutorService executor) throws Exception {
+        Field eField = Class.forName("java.util.concurrent.Executors$DelegatedExecutorService")
+                .getDeclaredField("e");
+        eField.setAccessible(true);
+        return (ExecutorService) eField.get(executor);
     }
 
-    //--------------------- Infrastructure ---------------------------
-    static volatile int passed = 0, failed = 0;
-    static void pass() {passed++;}
-    static void fail() {failed++; Thread.dumpStack();}
-    static void fail(String msg) {System.out.println(msg); fail();}
-    static void unexpected(Throwable t) {failed++; t.printStackTrace();}
-    static void equal(Object x, Object y) {
-        if (x == null ? y == null : x.equals(y)) pass();
-        else fail(x + " not equal to " + y);}
-    public static void main(String[] args) throws Throwable {
-        try {realMain(args);} catch (Throwable t) {unexpected(t);}
-        System.out.printf("%nPassed = %d, failed = %d%n%n", passed, failed);
-        if (failed > 0) throw new AssertionError("Some tests failed");}
+    /**
+     * Invokes System.gc and and waits for the given ExecutorService to terminate.
+     */
+    private void gcAndAwaitTermination(ExecutorService executor) throws Exception {
+        System.err.println(executor);
+        boolean terminated = false;
+        while (!terminated) {
+            System.gc();
+            terminated = executor.awaitTermination(100, TimeUnit.MILLISECONDS);
```

Comment:
@AlanBateman Perhaps worth having some upper limit as to how long it will wait until failing the test?

Reply:
jtreg will look after this. The default timeout is 120s but it can be adjusted with a timeout factor (make time uses a timeout factor of 4). So it would be possible to put a limit on this in the test but it can be problematic, e.g. with debug builds + Xcomp for example.

Reply:
I have written such code many times, and I've always had success with a (jtreg-scaled) 10 second timeout.  E.g.

static final long LONG_DELAY_MS = Utils.adjustTimeout(10_000);

(But that's before I became aware of existence of (slow) Raspberry Pis.)

I continue to feel guilty about never having written a proper gc awaiting library utility for openjdk, like I did for guava here:

https://github.com/google/guava/blob/master/guava-testlib/src/com/google/common/testing/GcFinalization.java

Reply:
Now there's [jdk.test.lib.util.ForceGC](https://github.com/openjdk/jdk/blob/master/test/lib/jdk/test/lib/util/ForceGC.java).